libtiledbsoma/src/utils/arrow_adapter.h

```diff
@@ -702,8 +702,61 @@ class ArrowAdapter {
         return arrow_array;
     }
 
-    // These table-column getters are, as of this writing, intended primarily
-    // for keystroke-reduction in unit-test cases.
+    // Given TileDB 8-bit booleans, packs them to 1-bit Arrow-style booleans.
+    static ArrowArray* make_arrow_array_child_bool(
+        const std::vector<uint8_t>& v) {
+        // Use malloc here, not new, to match ArrowAdapter::release_array
+        auto arrow_array = (ArrowArray*)malloc(sizeof(ArrowArray));
+
+        size_t n = v.size();
+
+        arrow_array->length = n;  // Number of bools, not number of bytes
+        arrow_array->null_count = 0;
+        arrow_array->offset = 0;
+
+        // Two-buffer model for non-string data:
+        // * Slot 0 is the Arrow validity buffer which we leave null
+        // * Slot 1 is data, void* but will be dereferenced as T*
+        // * There is no offset information
+        arrow_array->n_buffers = 2;
+        arrow_array->n_children = 0;  // leaf/child node
+        arrow_array->buffers = (const void**)malloc(2 * sizeof(void*));
+        arrow_array->children = nullptr;  // leaf/child node
+        arrow_array->dictionary = nullptr;
+
+        // The nominal use of these methods as of this writing is for
+        // low-volume data such as schema information -- less than a
+        // kilobyte total. It's simplest and safest to do data copies,
+        // for-loop-wise. If we were to extend usage of these methods
+        // to bulk data in the megabyte/gigabyte range, we'd want to
+        // look at zero-copy for buffers, with variable approaches
+        // to memory management.
+        arrow_array->release = &ArrowAdapter::release_array;
+        arrow_array->private_data = nullptr;
+
+        arrow_array->buffers[0] = nullptr;
+        // Use malloc here, not new, to match ArrowAdapter::release_array.
+        // When computing the number of bits needed to store the n bits, round
+        // up.
+        auto nbytes = (n + 7) / 8;
+        uint8_t* dest = (uint8_t*)malloc(nbytes * sizeof(uint8_t));
+        memset(dest, 0, nbytes);
+
+        for (size_t i = 0; i < n; i++) {
```

Comment:
I believe this is redundant with the nanoarrow routines which do the same thing [see here](https://arrow.apache.org/nanoarrow/0.2.0/c.html#bitmap-utilities)

Reply:
(I saw the "Unsafe" in that function name earlier today and quickly icked away from using it. I am writing code which I don't feel deserves worthy of being called "unsafe". I am happy with my work product, and am unit-testing it to my satisfaction. I'm putting up a PR for review here, for the express purpose of seeing if anyone finds it in need of further testing.)

Reply:
unsafe just means you need to do your own bounds checks and allocations.

the code is quite straightforward if you want to have a read (in nanoarrow.h)

Reply:
See also https://github.com/single-cell-data/TileDB-SOMA/pull/3857#discussion_r2016596989